### PR TITLE
Adding reduced Gsf tracks to miniAOD

### PR DESCRIPTION
**Pull request by andreypz**
This commit is to add reduced gsf tracks collection to miniAOD.
The tracks added are:
 * The main track of GsfElectron
 * All ambiguous tacks of GsfElectron (if there are any)

The intent  for this code to be used for miniAOD of 2017 data/MC and legacy re-miniAOD of 2016 data.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

@andreypz 
please provide file size change estimates due to proposed changes.
Also, if this was already presented in XPF, please add a link to the slides.

File: PhysicsTools/PatAlgos/runMiniAOD_90X.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
+# Auto generated configuration file
```
Comment (slava77) [inline PhysicsTools/PatAlgos/runMiniAOD_90X.py]: please remove this file:
it should be either in /test directory or not present at all
miniAOD processing is exercised already in many matrix workflows

File: RecoEgamma/EgammaPhotonProducers/interface/ReducedEGProducer.h [hunk 2]
```diff
@@ -90,18 +94,23 @@ class ReducedEGProducer : public edm::stream::EDProducer<> {
  const std::string outESRecHits_;
  const std::string outPhotonPfCandMap_;
  const std::string outGsfElectronPfCandMap_;
+ const std::string outGsfTracks_;  //add by me - Hien xinh dep
+ //const std::string outGsfTrackCores_; //add by me - Hien xinh dep
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/interface/ReducedEGProducer.h]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

if you need to keep these commants, please cleanup to be clear who "me" is ; I think this is not needed at all

File: RecoEgamma/EgammaPhotonProducers/interface/ReducedEGProducer.h [hunk 3]
```diff
@@ -90,18 +94,23 @@ class ReducedEGProducer : public edm::stream::EDProducer<> {
...
  const StringCutObjectSelector<reco::Photon> relinkPhotonSel_;
  const StringCutObjectSelector<reco::GsfElectron> keepGsfElectronSel_;
  const StringCutObjectSelector<reco::GsfElectron> slimRelinkGsfElectronSel_;
- const StringCutObjectSelector<reco::GsfElectron> relinkGsfElectronSel_; 
+ const StringCutObjectSelector<reco::GsfElectron> relinkGsfElectronSel_;
+
+ //add by Hien
+ // const StringCutObjectSelector<reco::GsfTrack> keepGsfTrackSel_;
+ //const StringCutObjectSelector<reco::GsfTrack> slimRelinkGsfTrackSel_;
+ //const StringCutObjectSelector<reco::GsfTrack> relinkGsfTrackSel_;
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/interface/ReducedEGProducer.h]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 4]
```diff
@@ -89,60 +91,61 @@ ReducedEGProducer::ReducedEGProducer(const edm::ParameterSet& config) :
...
-  const std::vector<edm::InputTag>& gsfelectronidinputs = 
+
+  const std::vector<edm::InputTag>& gsfelectronidinputs =
     config.getParameter<std::vector<edm::InputTag> >("gsfElectronIDSources");
   for (const edm::InputTag &tag : gsfelectronidinputs) {
     gsfElectronIdTs_.emplace_back(consumes<edm::ValueMap<float> >(tag));
-  }  
-  
-  const std::vector<edm::InputTag>&  photonpfclusterisoinputs = 
+  }
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: was there a good reason to make all these empty line/space changes?

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 5]
```diff
@@ -311,65 +321,104 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
...
+      gsfTrackMap[gsfTrack] = gsfTracks->size() - 1;
+      //std::cout << "push back gsftrack to gsftrack ref" << std::endl;
+      //std::cout << " gsf track size: " << gsfTracks->size() << std::endl;
+    }
+
+
+    //std::cout << " ---- event number: " << theEvent.id().event() << "\t gsf pt: " << gsfTrack->pt() << "\t gsf eta: "
+    //	      << gsfTrack->eta() << "\t gsf phi: " << gsfTrack->phi() << std::endl;
+    //std::cout << "ambiguous true or false: " <<gsfElectron.ambiguous()
+    //	      << "  (looks like this flag was never set even when there are ambiguous tracks present)" <<std::endl;
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

**Comment by slava77** [comment]
@cmsbuild please abort

**Comment by slava77** [comment]
-1

please follow the code review and resubmit this to /master.
There are no current plans for new 90X releases

**Comment by andreypz** [comment]
@slava77 
Thank you for the review and suggestions. I'm going to implement them by the end of the week.
Just one question, to confirm. Do you really mean that we need to make a pull request to master branch? I see that other requests here are made to 92X or 93X branches.

**Comment by slava77** [comment]
On 7/11/17 12:50 PM, Andrey Pozdnyakov wrote:
> @slava77 <https://github.com/slava77>
> Thank you for the review and suggestions. I'm going to implement them by
> the end of the week.
> Just one question, to confirm. Do you really mean that we need to make a
> pull request to master branch? I see that other requests here are made
> to 92X or 93X branches.

All new developments should start with a pull request to the master branch
(which is currently equivalent to CMSSW_9_3_X).
After that, a backport can be made if there is a good reason for that.

**Comment by andreypz** [comment]
Closing this one. 
Made another pull to master.

**Pull request closed without merging**